### PR TITLE
Variational autoencoder for tract profile data 

### PR DESCRIPTION
**Pull request by SamChou05**

### CODE REVIEW

**Comment by arokem** [comment]
@SamChou05 : now that #19 is merged, please go ahead and merge the main branch into your branch or rebase on top of main so that we can see whether your tests will pass.

File: afqinsight/nn/utils.py [hunk 1]
```diff
@@ -427,10 +429,107 @@ def prep_pytorch_data(dataset):
...
+
+
+def prep_fa_dataset(dataset, batch_size=32):
+    """
+    Extracts the fa_tracts from the provided dataset and
+    prepares the dataset for training.
+    """
+    fa_indices = []
+    for i, fname in enumerate(dataset.feature_names):
+        if fname[0] == "dki_fa":
```
Comment (arokem) [inline afqinsight/nn/utils.py]: This label may vary between datasets

File: afqinsight/nn/pt_models.py [hunk 2]
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
     return cnn_resnet_Model
 
 
+class VariationalEncoder(nn.Module):
+    def __init__(self, input_shape, latent_dims=20, dropout=0.2):
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: Looks like the `__init__` for this and the subsequent `Encoder` class have a lot in common. Are they actually identical, except for variable names? If so, maybe they can both rely on the same base class? Or one be the base and the other inherits from it?

File: afqinsight/nn/pt_models.py [hunk 3]
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
...
+        x = self.dropout(x)
+        x = self.relu(self.conv2(x))
+        x = self.dropout(x)
+        x = self.relu(self.conv3(x))
+        x = self.dropout(x)
+
+        # mu = self.conv_mu(x)
+        # logvar = self.conv_logvar(x)
+
+        x = self.flatten(x)  # [64, 64*7]
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: What does this comment mean?

File: afqinsight/nn/pt_models.py [hunk 4]
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
...
+
+    def forward(self, x):
+        x = self.relu(self.conv1(x))
+        x = self.dropout(x)
+        x = self.relu(self.conv2(x))
+        x = self.dropout(x)
+        x = self.relu(self.conv3(x))
+        x = self.dropout(x)
+
+        # mu = self.conv_mu(x)
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: Remove?

File: afqinsight/nn/pt_models.py [hunk 5]
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
...
+            32, 16, kernel_size=4, stride=2, padding=2, output_padding=1
+        )
+        self.deconv4 = nn.ConvTranspose1d(
+            16, 1, kernel_size=3, stride=2, padding=2, output_padding=1
+        )
+
+        self.relu = nn.ReLU()
+
+    def forward(self, x):
+        # x = F.relu(self.deconv1(x))
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: Remove?

File: afqinsight/nn/pt_models.py [hunk 6]
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
...
+        self.linear2 = nn.Linear(50, input_shape)
+
+    def forward(self, z):
+        x = self.linear1(z)
+        x = self.relu(x)
+        x = self.linear2(x)
+        return x
+
+
+class Conv1DVariationalEncoder(nn.Module):
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: Same comment as above. Could this be consolidated with the Conv1DEncoder?

File: afqinsight/nn/pt_models.py [hunk 7]
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
...
+        opt = torch.optim.Adam(self.parameters(), lr=lr)
+        scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
+            opt, "min", patience=5, factor=0.5
+        )
+
+        train_rmse_per_epoch = []
+        train_kl_per_epoch = []
+        train_recon_per_epoch = []
+
+        beta_start = 0.0
```
Comment (arokem) [inline afqinsight/nn/pt_models.py:745]:
```suggestion
        # This is the weight for the KL "annealing"
        beta_start = 0.0
```

File: afqinsight/nn/pt_models.py [hunk 8]
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
...
+        slope = (beta_end - beta_start) / epochs
+
+        for epoch in range(epochs):
+            running_loss = 0.0
+            running_rmse = 0.0
+            running_kl = 0.0
+            running_recon_loss = 0.0
+            items = 0
+
+            beta = beta_start + slope * epoch
```
Comment (arokem) [inline afqinsight/nn/pt_models.py:757]:
```suggestion
            # The weight on KL changes linearly with epoch:
            beta = beta_start + slope * epoch
```

File: afqinsight/nn/pt_models.py [hunk 9]
```diff
@@ -542,6 +547,624 @@ def cnn_resnet_pt(input_shape, n_classes):
...
+            scheduler.step(running_loss / items)
+            avg_train_rmse = running_rmse / items
+            avg_train_kl = running_kl / items
+            avg_train_recon_loss = running_recon_loss / items
+
+            train_rmse_per_epoch.append(avg_train_rmse)
+            train_kl_per_epoch.append(avg_train_kl)
+            train_recon_per_epoch.append(avg_train_recon_loss)
+
+            print(
```
Comment (arokem) [inline afqinsight/nn/pt_models.py]: Maybe use `logging` instead of `print`?

File: afqinsight/nn/pt_models.py (hunk 3 above)
Comment (SamChou05) [reply afqinsight/nn/pt_models.py]: It was a note for the shape of the data after flattening in one of the experiments I was running, getting rid of it as it shouldn't be there.

**Comment by arokem** [comment]
The code in pt_models looks great! 

I wonder whether `prep_first_tract_data`, `prep_fa_dataset`, and `prep_fa_flattned_data` are all a bit specific to the analysis that you are currently doing, so maybe they belong in your analysis code repo instead of here?

**Comment by SamChou05** [comment]
I agree with your comment with prep_first_tract_data, prep_fa_dataset, and prep_fa_flattned_data, so I moved them to my repository containing the experiments and removed them from this PR.

**Comment by arokem** [comment]
Great! You'll need to move out the related test as well

**Comment by arokem** [comment]
Could you please add docstrings to all the new functions/classes?
